core/commands/refs.go

```diff
@@ -274,43 +290,40 @@ func (rw *RefWriter) writeRefsRecursive(n ipld.Node) (int, error) {
 	return count, nil
 }
 
-func (rw *RefWriter) writeRefsSingle(n ipld.Node) (int, error) {
-	c := n.Cid()
-
-	if rw.skip(c) {
-		return 0, nil
+// visit returns false if we should not re-explore the tree under a Cid, or
+// true otherwise. The second return argument indicates whether the Cid was seen
+// before.
+func (rw *RefWriter) visit(c *cid.Cid, depth int) (bool, bool) {
+	if rw.seen == nil {
```

Comment:
I believe disabling unique by default is a memory optimization. If it's disabled, we should avoid using memory linear in the number of keys. We should probably have check for the `Unique` flag up-top and, in that case, only check the depth (don't store anything).

Reply:
It may also be simpler to have one boolean indicate that we should continue traversing and the other indicate that we should return the CID to the user. That may simplify some of the other logic. It'll also allow us to say "return this but *don't* traverse it which, unless I'm mistaken, should save us a bit of work (possibly saving us from fetching a node we don't need to traverse).